pkg/generator.go

```diff
@@ -611,26 +611,45 @@ func (g *Generator) Generate(ctx context.Context) error {
 			ret := make([]string, len(returns.Types))
 
 			for idx, typ := range returns.Types {
+				ret[idx] = fmt.Sprintf("r%d", idx)
 				g.printf("\tvar r%d %s\n", idx, typ)
-				g.printf("\tif rf, ok := %s.Get(%d).(func(%s) %s); ok {\n",
-					retVariable, idx, strings.Join(params.Types, ", "), typ)
-				g.printf("\t\tr%d = rf(%s)\n", idx, formattedParamNames)
+			}
+
+			var tab string
+			if len(returns.Types) > 1 {
+				tab = "\t"
+				g.printf("\tif rf, ok := %s.Get(0).(func(%s) (%s)); ok {\n",
+					retVariable, strings.Join(params.Types, ", "), strings.Join(returns.Types, ", "))
+				g.printf("\t\t%s = rf(%s)\n", strings.Join(ret, ", "), formattedParamNames)
 				g.printf("\t} else {\n")
+			}
+
+			for idx, typ := range returns.Types {
+				if idx > 0 {
+					g.printf("\n")
+				}
+
+				g.printf("%s\tif rf, ok := %s.Get(%d).(func(%s) %s); ok {\n",
+					tab, retVariable, idx, strings.Join(params.Types, ", "), typ)
+				g.printf("%s\t\tr%d = rf(%s)\n", tab, idx, formattedParamNames)
+				g.printf("%s\t} else {\n", tab)
```

Comment:
You know a part of me wants to stop using individual printf statements because it's difficult for me to understand what's going on. You have the right approach of trying to follow what the code historically did, but I think it'd be better for readability to use the `printTemplate` method instead. What do you think? Is that easily done or is printf a better fit here?

Reply:
I don't find either of them all that easy to follow, tbh.  It's easier to look at the expected outputs in generator_test.go, and trust the tests to make sure all the templates and printf's are actually producing that.

Reply:
Comparing before and after on one of the test cases, this PR changes this:

```go
// Put provides a mock function with given fields: path
func (_m *RequesterReturnElided) Put(path string) (int, error) {
	ret := _m.Called(path)

	var r0 int
	if rf, ok := ret.Get(0).(func(string) int); ok {
		r0 = rf(path)
	} else {
		r0 = ret.Get(0).(int)
	}

	var r1 error
	if rf, ok := ret.Get(1).(func(string) error); ok {
		r1 = rf(path)
	} else {
		r1 = ret.Error(1)
	}

	return r0, r1
}
```

into this:

```go
// Put provides a mock function with given fields: path
func (_m *RequesterReturnElided) Put(path string) (int, error) {
	ret := _m.Called(path)

	var r0 int
	var r1 error
	if rf, ok := ret.Get(0).(func(string) (int, error)); ok {
		r0, r1 = rf(path)
	} else {
		if rf, ok := ret.Get(0).(func(string) int); ok {
			r0 = rf(path)
		} else {
			r0 = ret.Get(0).(int)
		}

		if rf, ok := ret.Get(1).(func(string) error); ok {
			r1 = rf(path)
		} else {
			r1 = ret.Error(1)
		}
	}

	return r0, r1
}
```

And adds this to the expecter calls:

```go
func (_c *RequesterReturnElided_Put_Call) RunAndReturn(run func(string) (int, error)) *RequesterReturnElided_Put_Call {
	_c.Call.Return(run)
	return _c
}
```

Reply:
I can try, but that's a heck of a lot more work to rewrite more of the code than what I've modified so far, and getting templates to produce exactly the same output might be pretty annoying.

Reply:
This was a _giant_ pain in the butt, and I did have to tweak the new output slightly, but I managed to get a template to generate the same code that the printf stuff was doing.  (Technically `generateCalled` is sort of cheating and just using `fmt.Sprintf` instead of `g.printf` in a few places, but it'll do for now.)

The thing that I changed in the expected outputs is to do this:

```go
	if rf, ok := ret.Get(0).(func(string, uint16) ([]byte, *test.Err)); ok {
		return rf(_a0, _a1)
	}
	if rf, ok := ret.Get(0).(func(string, uint16) []byte); ok {
		r0 = rf(_a0, _a1)
	} else {
		if ret.Get(0) != nil {
			r0 = ret.Get(0).([]byte)
		}
	}

	if rf, ok := ret.Get(1).(func(string, uint16) *test.Err); ok {
		r1 = rf(_a0, _a1)
	} else {
		if ret.Get(1) != nil {
			r1 = ret.Get(1).(*test.Err)
		}
	}

	return r0, r1
```

Note the early return if the new type of provider function is found.  This way there doesn't need to be an `else` block, and the indentation of the rest of the function is much easier to deal with.

Reply:
(Don't even ask me what's going on with all the `{{- }}` and `{{ -}}` bits in the new template parts.  I was just poking at it until all the whitespace exactly matched what was already in the test expectations.

Reply:
This is great! Sorry, if I knew how much of a pain it would be maybe I could have let it slide but I do really enjoy the templates a lot more. At least to me, it's easier to reason about.

I'll ack this, lmk if there are any finishing touches you want to make before I merge.

Reply:
I need to walk away for a bit and look at it with fresh eyes later, maybe in the morning.  I'll let you know if I spot anything that could use work then.

Reply:
Sounds good, thanks for the stellar work. It's really appreciated.

Reply:
I think this is good to go.  There was one function comment that was sort of inaccurate after I changed things, but I think the code itself is okay.

Reply:
Great work, I’ll take one last look tonight and merge if everything is good. Thanks Dave.